streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareTaskAssignor.java

```diff
@@ -38,29 +43,34 @@
 
 public class RackAwareTaskAssignor {
     private static final Logger log = LoggerFactory.getLogger(RackAwareTaskAssignor.class);
+    private static final int DEFAULT_STATEFUL_TRAFFIC_COST = 10;
```

Comment:
Why is this set to `10`? In particular, why is it higher than for the stateless case? In the end, my understanding was that we try to optimize for input partitions, and for this case, there is no difference if a task has state or not, but only the number of input topic partitions for a task matter (each with equal cost)

Reply:
For stateless tasks, there's no `non_overlap_cost`, so traffic cost doesn't matter as long as it's positive. The reason is that it's less expensive to relocate stateless tasks.

For stateful tasks, there's `non_overlap_cost` which means there's cost if you are moving the task to other clients. Since we prefer not to move stateful tasks as much as possible, there's non-zero value for this cost. The value for `traffic_cost` vs `non_overlap_cost` means how much we favor one compared to another. Setting to 10 basically means we value `traffic_cost` much more than `non_overlap_cost`. Other default could also be possible. Also, since this class is also going to be used by `StickyAssignor`, my plan is to assign `non_overlap_cost` higher value than `traffic_cost` for sticky assignor to favor stickiness much more

Reply:
One thing missing here is that I intend to have different default stateful traffic cost and non-overlap cost for `StickyAssignor` and `HAAssignor`, so instead of hardcoding here, maybe it makes more sense to pass in from constructor or assign() call.

This brings to another question of where should `RackAwareTaskAssignor` be constructed. Before `StickyAssignor` and `HAAssignor` are constructed using reflection, we can't pass `RackAwareTaskAssignor` to them via their constructor. We can construct `RackAwareTaskAssignor` inside them, but constructor of `RackAwareTaskAssignor` requires a lot more params than `HAAssignor`'s assign() call. So I'm thinking construct `RackAwareTaskAssignor` inside `StreamsPartitionAssignor` and pass it to `HAAssignor` or `StickyAssignor` using `assignmentConfig` variable...